dsse/verify.go

```diff
@@ -1,70 +1,116 @@
-/*
-Package dsse implements the Dead Simple Signing Envelope (DSSE)
-https://github.com/secure-systems-lab/dsse
-*/
 package dsse
 
+import (
+	"crypto"
+	"errors"
+	"fmt"
+)
+
 /*
 Verifier verifies a complete message against a signature and key.
 If the message was hashed prior to signature generation, the verifier
 must perform the same steps.
-If the key is not recognized ErrUnknownKey shall be returned.
+If KeyID returns successfully, only signature matching the key ID will be verified.
 */
 type Verifier interface {
-	Verify(keyID string, data, sig []byte) error
+	Verify(data, sig []byte) error
+	KeyID() (string, error)
+	Public() crypto.PublicKey
 }
 
-type EnvelopeVerifier struct {
+type envelopeVerifier struct {
 	providers []Verifier
+	threshold int
 }
 
-func (ev *EnvelopeVerifier) Verify(e *Envelope) error {
+type AcceptedKeys struct {
+	Public crypto.PublicKey
+	KeyID  string
+	Sig    Signature
+}
+
+func (ev *envelopeVerifier) Verify(e *Envelope) ([]AcceptedKeys, error) {
 	if len(e.Signatures) == 0 {
-		return ErrNoSignature
+		return nil, ErrNoSignature
 	}
 
 	// Decode payload (i.e serialized body)
 	body, err := b64Decode(e.Payload)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	// Generate PAE(payloadtype, serialized body)
 	paeEnc := PAE(e.PayloadType, string(body))
 
-	// If *any* signature is found to be incorrect, the entire verification
-	// step fails even if *some* signatures are correct.
-	verified := false
+	// If *any* signature is found to be incorrect, it is skipped
+	var acceptedKeys []AcceptedKeys
+	usedKeyids := make(map[string]string)
 	for _, s := range e.Signatures {
 		sig, err := b64Decode(s.Sig)
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		// Loop over the providers. If a provider recognizes the key, we exit
+		// Loop over the providers.
+		// If provider and signature include key IDs but do not match skip.
+		// If a provider recognizes the key, we exit
 		// the loop and use the result.
 		for _, v := range ev.providers {
-			err := v.Verify(s.KeyID, paeEnc, sig)
+			keyID, err := v.KeyID()
+			if s.KeyID != "" && keyID != "" && err == nil && s.KeyID != keyID {
+				continue
+			}
 			if err != nil {
-				if err == ErrUnknownKey {
-					continue
-				}
-				return err
+				keyID = ""
+			}
+
+			err = v.Verify(paeEnc, sig)
+			if err != nil {
+				continue
+			}
+
+			acceptedKey := AcceptedKeys{
+				Public: v.Public(),
+				KeyID:  keyID,
+				Sig:    s,
 			}
 
-			verified = true
+			// See https://github.com/in-toto/in-toto/pull/251
+			if val, ok := usedKeyids[keyID]; ok {
+				fmt.Printf("Found envelope signed by different subkeys of the same main key, Only one of them is counted towards the step threshold, KeyID=%s\n", val)
```

Comment:
I think this needs a continue statement as well so that the key is not added to `acceptedKeys`? Or at least a test to make sure it skips the duplicate.